sdk/core/core/src/az_log.c

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_log.h>
+
+#include <_az_cfg.h>
+
+static az_log_classification const * _az_log_classifications = NULL;
+static size_t _az_log_classifications_length = 0;
+static az_log * _az_log_listener = NULL;
+
+void az_log_set_classifications(
+    az_log_classification const * const classifications,
+    size_t const classifications_length) {
+  _az_log_classifications = classifications;
+  _az_log_classifications_length = classifications_length;
+}
+
+void az_log_reset_classifications() {
+  _az_log_classifications = NULL;
+  _az_log_classifications_length = 0;
+}
+
+void az_log_set_listener(az_log * const listener) { _az_log_listener = listener; }
+
+void az_log_write(az_log_classification const classification, char const * const message) {
+  if (_az_log_listener != NULL && az_log_should_write(classification)) {
+    (*_az_log_listener)(classification, message);
+  }
+}
+
+bool az_log_should_write(az_log_classification const classification) {
+  if (_az_log_classifications == NULL || _az_log_classifications_length == 0) {
```

Comment:
Please add a comment about the intent here: "If the user hasn't registered any classifications, then we log everything."
Now that I think about it, we may not want this behavior. What should default logging be? Maybe nothing. I'll talk to the IoT team about this.

Reply:
I added the comment. Maybe it is fine to log everything by default, as opposed to nothing.
1. In case user wants no log, they don't set the listener.
2. In case user wants everything, they don't specify the classifications. So, when we start logging new types of classifications, user starts to get them as well.

Ideally, for the maximum flexibility, there should be a way to include and exclude classifications, i.e. "I want everything except the slow response errors". And then when we add new types of classifications, user starts getting them as well. Or, if user told us they want only the retry and error messages, they get only these types of log entries all the time, and nothing more. Or, user may want to only get the keyvault errors, and nothing else.

For now, I think, our approach is fine, until we know of better scenarios.
But still, did you have a chance to talk to the IoT team about this?